include/vcpkg/base/parallel-algorithms.h

```diff
@@ -0,0 +1,24 @@
+#pragma once
+
+#if defined(USE_PARALLEL_ALG) && !defined(GNU_USE_PARALLEL_ALG)
+#include <algorithm>
+#include <execution>
+#elif defined(GNU_USE_PARALLEL_ALG)
+#include <parallel/algorithm>
+#else
+#include <algorithm>
+#endif
+
+#if defined(USE_PARALLEL_ALG) && !defined(GNU_USE_PARALLEL_ALG)
+#define vcpkg_parallel_for_each std::for_each(std::execution::par,
+#elif defined(GNU_USE_PARALLEL_ALG)
+#define vcpkg_parallel_for_each __gnu_parallel::for_each(
```

Comment:
https://gcc.gnu.org/onlinedocs/libstdc++/manual/parallel_mode.html:
> For C++17 and above there are new overloads of the standard algorithms which take an execution policy argument. You should consider using those instead of the non-standard parallel mode extensions. 

Do I understand correctly that `__gnu_parallel::for_each` shouldn't be used? 

I dislike the macro stuff here. And the unbalanced `(` in definition and `)` when used are not acceptable IMO.

Reply:
> Do I understand correctly that `__gnu_parallel::for_each` shouldn't be used?

> For C++17 and above there are new overloads of the standard algorithms which take an execution policy argument. You should consider using those instead of the non-standard parallel mode extensions.

I considered using STL parallel algorithms but they need tbb as a dependence which is not available in CI and on people's devices. I already discussed this with BillyONeal on Discord: We can't depend on tbb. Therefore, I think I did the right think here since `__gnu_parallel` is not deprecated.

> I dislike the macro stuff here. And the unbalanced ( in definition and ) when used are not acceptable IMO.

As stated above, macros are the fastest solution here als regular functions can't be inlined. I definately agree  on the braces with you. I will work out a better solution here.

Reply:
> As stated above, macros are the fastest solution here als regular functions can't be inlined.

The compiler might even inline a regular function if optimization is enabled. But we are talking about template functions here, not regular functions. Macros may be faster at compile time, but I don't think there is an advantage at run time.

Reply:
> The compiler might even inline a regular function if optimization is enabled. But we are talking about template functions here, not regular functions. Macros may be faster at compile time, but I don't think there is an advantage at run time.

I have already benchmarked that using the template function is slower than using `std:.for_each` directly with optimization turned on. To repeat: The templated function doesn't get inlined, which makes it slower. I already have benchmarked this. If you don't believe me, feel free to benchmark this on your own. If you are not able to provide a suggestion that is faster than my code, I won't change this.

IMO there is nothing wrong with using macros as functions.